### PR TITLE
Feature/webhook admin

### LINKED ISSUE #115: Refresh Webhook
When content gets sunrise and sunset, we should provide a way to reach out to an application and let them know they need to refresh their content

```gherkin
Feature: Register Application
  As an administrator
  I want to register endpoint to call when content has changed
  So that application developers can handle content changes automatically

  Scenario: Register
    Given that I know the URLs of endpoints to call on content actions
     When I navigate to /applications/new
     Then I can add a new application
      And I can set the application's name, which is required
      And I can set an endpoint to call when content goes live
      And I can set an endpoint to call when live content gets updated
      And I can set an endpoint to call when live content gets sunset 
      And I will receive a randomly generated, unique, hard to guess Client ID
      And I will receive a randomly generated, unique, hard to guess Client Secret
      And it will be assigned an internal ID
      And I will be redirected to its individual view page once saved

  Scenario: View All
    Given that I have one or more applications registered
     When I navigate to /applications
     Then I can see a list of all of the applications registered
      And I can see their names
      And I can see if they have a go live endpoint registered
      And I can see if they have an update endpoint registered
      And I can see if they have a sunset endpoint registered
      And I can see a black space if an endpoint is not registered
      And I can see an indicator if the the endpoint was successfully called or not if an endpoint was registered
      And the indicator will link to the edit view of the application showing history of that endpoint's calls
      And I will have a link to edit the application
      And the list of applications will be paginated in groups of 10

  Scenario: Add New
    Given that I wish to register an application
     When I navigate to /applications
     Then I click on a button that will allow me to add a new application
      And if there is a list of registered applications, the button will be above them

  Scenario: View One
    Given that I have an application registered
     When I navigate to /applications/:id
     Then I can see the application's name
      And I can see each of the application's endpoints
      And both the Client ID and Client Secret will be displayed
      And I can reset the Client Secret
      And I can update the application's name or endpoints
      And I can see the call history of each endpoint which includes the time it was run and if the call was successful or not
      And I can delete the application


```

---

**Register Webhook**
- [x] Register
- [x] View All
- [x] Add New
- [x] View One

### PR DESCRIPTION
**Pull request by scottnath**
@Snugug @joshblack 

This is the final PR for creating webhooks management in Punchcard. It sure is big! Good thing my tests didn't need changes except pulling out the `skip`s ;-)

---
Resolves #115 

`DCO 1.1 Signed-off-by: Scott Nath <github@scottnath.com>`

### CODE REVIEW

File: lib/applications/routes.js [hunk 1]
```diff
@@ -4,53 +4,322 @@
...
  * @param {object} res - HTTP Response
  * @param {object} next - Express callback
  *
+
+ * @returns {promise} promise wrapping database promise
  */
-const all = () => {};
+const all = (req, res, next) => {
+  return new Promise((resolve, reject) => {
+    return database
```
Comment (Snugug) [inline lib/applications/routes.js]: Database is a promise, don't wrap it in a promise

File: lib/applications/routes.js [hunk 2]
```diff
@@ -4,53 +4,322 @@
...
+      .select('*')
+      .from(`${config.applications.base}`)
```
Comment (Snugug) [inline lib/applications/routes.js]: Our Database table names aren't configurable. This is [always going to be `applications`](https://github.com/punchcard-cms/punchcard/blob/master/lib/database/schemas.js#L238-L239). This is an incorrect use of configuration

File: lib/applications/routes.js [hunk 3]
```diff
@@ -4,53 +4,322 @@
...
+      .orderBy('name', 'DESC').then(rws => {
+        // add itentifier to each row
+        const rows = utils.routes.identifier(rws, req.applications.merged);
```
Comment (Snugug) [inline lib/applications/routes.js]: This is _not_ our content. Why are you running this through the route identifier instead of just using the `name` column readily available in the rows you're going to be retrieving?

File: lib/applications/routes.js [hunk 4]
```diff
@@ -4,53 +4,322 @@
...
  * https://github.com/punchcard-cms/content-types/issues/112
  */
-const add = () => {};
+const add = (req, res, next) => {
+  const errors = _.get(req.session, 'form.applications.save.errors', {});
+  const values = _.get(req.session, 'form.applications.save.content', {});
+
+  _.unset(req.session, 'form.applications.save');
+
+  return types.only(config.applications.base, values, [req.applications.merged], config).then(merged => {
```
Comment (Snugug) [inline lib/applications/routes.js]: The `base` config _is not_ a reliable source for us to use for things like database tabes and content configuration. While that _may_ hold true, now, those need to be hard-coded values because if that changes in config, everything will break.

File: lib/applications/routes.js [hunk 5]
```diff
@@ -4,53 +4,322 @@
...
+          return row;
+        });
+
+        res.render('applications/all', {
+          applications,
+          config: config.applications,
+          structure: model.structure,
+        });
+
+        resolve(true);
```
Comment (Snugug) [inline lib/applications/routes.js]: Should `return true` for a success

File: lib/applications/routes.js [hunk 6]
```diff
@@ -4,53 +4,322 @@
...
+      }).catch(e => {
+        const err = {
+          message: e.message,
+          safe: '/',
+          status: 500,
+        };
+
+        reject(err);
```
Comment (Snugug) [inline lib/applications/routes.js]: Should `return next(err)` instead of rejecting

File: lib/applications/routes.js [hunk 7]
```diff
@@ -4,53 +4,322 @@
...
  * View/edit an individual Webhook
  * @param {object} req - HTTP Request
  * @param {object} res - HTTP Response
  * @param {object} next - Express callback
  *
+ * @returns {promise} resolves/rejects promise
  */
-const one = () => {};
+const one = (req, res, next) => {
+  const id = _.get(req.session, 'form.applications.edit.id', null);
```
Comment (Snugug) [inline lib/applications/routes.js]: Won't `id` come from `req.params.id`?

File: lib/applications/routes.js [hunk 8]
```diff
@@ -4,53 +4,322 @@
...
+      safe: `/${config.applications.base}`,
+      status: 500,
+    };
+
+    return next(err);
+  });
+};
 
 /**
  * View/edit an individual Webhook
```
Comment (Snugug) [inline lib/applications/routes.js]: Individual Application, not Webhook

File: lib/applications/routes.js [hunk 9]
```diff
@@ -4,53 +4,322 @@
...
+  return new Promise((resolve, reject) => {
+    // something went wrong on save:
+    if (Object.keys(values).length > 0 && id) {
+      // add the previous session data back in
+      _.set(req.session, 'form.applications.edit', {
+        id,
+      });
+
+      // grab the edit form and inject errors and values from req.session
+      return types.only(config.applications.base, values, [req.applications.merged], config).then(merged => {
```
Comment (Snugug) [inline lib/applications/routes.js]: Same `base` issue here

File: lib/applications/routes.js [hunk 10]
```diff
@@ -4,53 +4,322 @@
...
+const one = (req, res, next) => {
+  const id = _.get(req.session, 'form.applications.edit.id', null);
+  const errors = _.get(req.session, 'form.applications.save.errors', {});
+  const values = _.get(req.session, 'form.applications.save.content', {});
+  let app = _.get(req.session, 'form.applications.edit.app', null);
+
+  _.unset(req.session, 'form.applications.edit');
+  _.unset(req.session, 'form.applications.save');
+
+  return new Promise((resolve, reject) => {
```
Comment (Snugug) [inline lib/applications/routes.js]: Same promise wrap here as above

File: lib/applications/routes.js [hunk 11]
```diff
@@ -4,53 +4,322 @@
...
-const secret = () => {};
+const secret = (req, res, next) => {
+  const id = _.get(req.session, 'form.applications.edit.id', null);
+  const referrer = req.get('Referrer');
+
+  const data = {
+    'client-secret': uuid.v4(),
+  };
+
+  return new Promise((resolve, reject) => {
```
Comment (Snugug) [inline lib/applications/routes.js]: Promise wrap

File: lib/applications/routes.js [hunk 12]
```diff
@@ -4,53 +4,322 @@
...
+      const err = {
+        message: 'Secret can only be changed from the application edit screen',
+        safe: `/${config.applications.base}`,
+        status: 500,
+      };
+
+      reject(err);
+    }
+
+    return database(`${config.applications.base}`).where('id', '=', id).update(data).then(() => {
```
Comment (Snugug) [inline lib/applications/routes.js]: Same database issue

File: lib/applications/routes.js [hunk 13]
```diff
@@ -4,53 +4,322 @@
...
+
+      reject(err);
+    });
+  }).catch(err => {
+    return next(err);
+  });
+};
 
 /**
  * View/edit an individual Webhook
```
Comment (Snugug) [inline lib/applications/routes.js]: This should be `Save an application`

File: lib/applications/routes.js [hunk 14]
```diff
@@ -4,53 +4,322 @@
...
+ * @returns {promise} database promise
  */
-const save = () => {};
+const save = (req, res) => {
+  const id = _.get(req.session, 'form.applications.edit.id', null);
+  const referrer = (_.get(req.session, 'referrer') || req.get('Referrer')) || `/${config.applications.base}`;
+
+  // user hit delete button
+  if (req.body.submit === config.applications.actions.delete && id) {
+    return database(`${config.applications.base}`).where('id', '=', id).del().then(() => {
```
Comment (Snugug) [inline lib/applications/routes.js]: Config database name

File: lib/applications/routes.js [hunk 15]
```diff
@@ -4,53 +4,322 @@
...
+
+        return res.redirect(`/${config.applications.base}`);
+      });
+    }
+
+    // on add, create fresh id/secret
+    data['client-id'] = uuid.v4();
+    data['client-secret'] = uuid.v4();
+
+    return database(`${config.applications.base}`).insert(data).returning('id').then((newid) => {
```
Comment (Snugug) [inline lib/applications/routes.js]: Config database name

File: views/applications/all.html [hunk 16]
```diff
@@ -12,7 +12,7 @@
   <div class="introduction">
     <h1 class="introduction--h1">{{structure.name}}</h1>
     <p class="introduction--p">{{structure.description}}</p>
-    <a href="/{{config.base}}/{{config.actions.add}}" class="introduction--button">{{config.actions.add|capitalize}} {{config.base}}</a>
+    <a href="/{{config.base}}/{{config.actions.add}}" class="introduction--button">{{config.actions.add|capitalize}} Application</a>
```
Comment (Snugug) [inline views/applications/all.html]: Now why isn't `config.base` being used here?

File: views/applications/one.html [hunk 17]
```diff
@@ -25,15 +26,16 @@ <h1 class="introduction--h1">{{config.actions.edit|capitalize}}</h1>
...
   {% endif %}
 
   {% if app.endpoints %}
     <h2>Endpoints</h2>
 
-    {% if app.endpoints.live %}
-      <h3 id="live">Live</h3>
+    {% if app.endpoints.live | length %}
+      <h3>Live</h3>
+      <a name="live"></a>
```
Comment (Snugug) [inline views/applications/one.html]: Why were these changed from `id` to `name`? `id` is semantically correct and `name` [has been deprecated](https://www.w3.org/TR/html5/obsolete.html) on anchor tags. Please return this to having `id` on the `h3`

**Comment by Snugug** [comment]
@scottnath Test are stalling at `applications › Single application route`

File: lib/applications/routes.js (hunk 7 above)
Comment (scottnath) [inline lib/applications/routes.js]: no, this id is specifically coming from session (or null). It is set in `save` and used here if save rejects back to `one`.

**Comment by scottnath** [comment]
@Snugug so, I've finally got all my tests except one to pass, which I skipped. Literally the only diff with the one before it (that passes) and this one ('Save new application') is that knex does an insert and return. I think this test should be put on hold until we can come up with a mock of the database for testing.

Unfortunately, now tests for `schedule` are failing every time. It's either the `schedule > push` or `schedule > pull` test (not both.) One of them fails every time. Can you fix that please so I can complete this PR? Or should I skip them?

**Comment by Snugug** [comment]
Skip them for now, they were passing (otherwise they wouldn't have found their way in) and we shouldn't block this PR on being able to track that down

File: lib/applications/routes.js [hunk 18]
```diff
@@ -4,53 +4,317 @@
...
+  const errors = _.get(req.session, 'form.applications.save.errors', {});
+  const values = _.get(req.session, 'form.applications.save.content', {});
+
+  _.unset(req.session, 'form.applications.save');
+
+  return types.only('applications', values, [req.applications.merged], config).then(merged => {
+    return types.form(merged, errors, config).then(form => {
+      res.render('applications/add', {
+        form,
+        action: req.url.replace(config.applications.actions.add, config.applications.actions.save),
```
Comment (Snugug) [inline lib/applications/routes.js]: This is what got us in to trouble with the content routes. Can we take the same tactic we did there for this action?

Comment (Snugug) [inline lib/applications/routes.js]: And you have below

File: lib/applications/routes.js (hunk 7 above)
Comment (Snugug) [inline lib/applications/routes.js]: OK

File: views/applications/_form.html [hunk 19]
```diff
@@ -0,0 +1,13 @@
+<form action="{{action}}" method="post" enctype="multipart/form-data" class="form" novalidate>
```
Comment (Snugug) [inline views/applications/_form.html]: Why this over the previously created [basic form](https://github.com/punchcard-cms/punchcard/blob/master/views/_basic-form.html)

File: tests/applications.js [hunk 20]
```diff
@@ -117,7 +117,7 @@ test('Workflow model from config', t => {
 //////////////////////////////
 // Routes - Applications landing
 //////////////////////////////
-test.cb.skip('All applications route', t => {
+test.cb.serial('All applications route', t => {
```
Comment (scottnath) [inline tests/applications.js]: I'll remove the serial. that was from testing to find the issues

File: views/applications/_form.html (hunk 19 above)
Comment (scottnath) [inline views/applications/_form.html]: should have been deleted. removing now.
